src/daft-logical-plan/src/ops/set_operations.rs

```diff
@@ -108,3 +110,96 @@ impl Intersect {
+    /// ex:
+    /// ```sql
+    /// > create table t0 as select 'hello' as c0;
+    /// > create table t1 as select 1 as c0;
+    /// > select * from t0 union select * from t1;
+    /// ```
+    /// This is valid in Union, but not in Concat
+    pub(crate) fn try_new(
+        lhs: Arc<LogicalPlan>,
+        rhs: Arc<LogicalPlan>,
+        is_all: bool,
+    ) -> logical_plan::Result<Self> {
+        if lhs.schema().len() != rhs.schema().len() {
+            return Err(DaftError::SchemaMismatch(format!(
+                "Both plans must have the same num of fields to union, \
+                but got[lhs: {} v.s rhs: {}], lhs schema: {}, rhs schema: {}",
+                lhs.schema().len(),
+                rhs.schema().len(),
+                lhs.schema(),
+                rhs.schema()
+            )))
+            .context(CreationSnafu);
+        }
+        Ok(Self { lhs, rhs, is_all })
+    }
+
+    /// union could be represented as a concat + distinct
+    /// while union all could be represented as a concat
+    pub(crate) fn to_logical_plan(&self) -> logical_plan::Result<LogicalPlan> {
+        let lhs_schema = self.lhs.schema();
+        let rhs_schema = self.rhs.schema();
+        let (lhs, rhs) = if lhs_schema != rhs_schema {
+            // we need to try to do a type coercion
+            let coerced_fields = lhs_schema
+                .fields
+                .values()
+                .zip(rhs_schema.fields.values())
+                .map(|(l, r)| {
+                    let new_dtype = get_supertype(&l.dtype, &r.dtype).ok_or_else(|| {
+                        logical_plan::Error::CreationError {
+                            source: DaftError::ComputeError(
+                                format!("
+                                    unable to find a common supertype for union. {} and {} have no common supertype",
+                                l.dtype, r.dtype
+                                ),
+                            ),
+                        }
+                    })?;
+                    Ok::<_, logical_plan::Error>(Field::new(l.name.clone(), new_dtype))
+                })
+                .collect::<Result<Vec<_>, _>>()?;
+            let projection_fields = coerced_fields
+                .into_iter()
+                .map(|f| col(f.name.clone()).cast(&f.dtype))
+                .collect::<Vec<_>>();
+            let lhs = Project::try_new(self.lhs.clone(), projection_fields.clone())?.into();
+            let rhs = Project::try_new(self.rhs.clone(), projection_fields)?.into();
+            (lhs, rhs)
+        } else {
+            (self.lhs.clone(), self.rhs.clone())
+        };
+        // we don't want to use `try_new` as we have already checked the schema
+        let concat = LogicalPlan::Concat(Concat {
+            input: lhs,
+            other: rhs,
+        });
+        if self.is_all {
+            Ok(concat)
+        } else {
+            Ok(Distinct::new(concat.arced()).into())
+        }
```

Comment:
Looks great in general, but I have a question: why not put this in the SQL planner instead of having a union operator at all? It seems like it should never actually be used, so is there a need to actually have this be a logical plan node type?

Reply:
(similar question for intersect but I did not get a chance to review that before it was merged)

Reply:
> Looks great in general, but I have a question: why not put this in the SQL planner or builder instead of having a union operator at all? It seems like it should never actually be used, so is there a need to actually have this be a logical plan node type?

i think it really comes down to separating out the logic, and future proofing ourselves. The sql planner already handles a lot of logic about how to convert a sql query to a plan, and having all of this additional logic kinda clutters the planner. Also, this seems like it's slightly different as a 'Union' and 'Intersect' are both valid logical operations, which we _may_ have optimizer rules for at some point. So doing it this way allows us to add in those optimizations when needed without having to do a larger scale refactor.

Reply:
Gotcha. I guess this is really about what the boundary should be between the logical plan and the things that construct logical plans. We can think about that in a different time then.